### DISCUSSION TITLE
Data Fetching in the app router with Axios and React Query

### DISCUSSION THREAD

**Discussion by FleetAdmiralJakob** [vercel/next.js#51091] [App Router] [answered]
If I want to use Axios and React Query to fetch my data how can I implement features like ISR in it?

**Comment by icyJoseph** [comment, marked as answer]
There's no inherent difference here, other than, during server side rendering, you won't get optimizations from the `fetch` server side version.

Read on here: https://nextjs.org/docs/app/building-your-application/data-fetching/fetching#segment-cache-configuration

That being said, today, there's little reason to use axios over fetch. You can even use https://www.npmjs.com/package/wretch if you find the fetch API  a bit awkward.

Regarding react-query, it is a client side fetching library, not sure how you image ISR fits in here, but you can always hydrate the client, https://tanstack.com/query/v4/docs/react/guides/ssr#using-the-app-directory-in-nextjs-13

**Reply by jakubjereczek** [reply]
@icyJoseph The main reason that Axios is used is it have built-in interceptors.

**Reply by icyJoseph** [reply]
Yeah, and a few other goodies, like automatic xsfr handling and whatnot.

The thing here is, should you go with that much from the start? It's not like any of that is not doable with simple fetch. 

- https://bundlephobia.com/package/axios

That's a lot of JavaScript. Surely as your app grows in complexity, you may find yourself writing a mini-axios within your codebase, at perhaps even higher JavaScript cost, then its best to take it. The right tool for the right job.

Funny that you mention interceptors because in projects where I've helped, that's one common source of bugs throughout the app.

**Comment by AndonMitev** [comment]
Hello, do you have any solutions for interceptors. I'm trying to use native fetch buuut having so hard times to create request interceptor and apply auth token OR, having error interceptor that listen for 401.

So far i have created something like this:

```
export const fetchApi = async (url: RequestInfo | URL, data?: any) => {
  const session = await getSession();

  const init = data
    ? {
        body: JSON.stringify(data),
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${session?.user.accessToken}`
        }
      }
    : undefined;

  const response = await fetch(url, { ...init });

  if (!response.ok) {
    if (response.status === 401) {
      throw new Error('Unauthorized');
    }

    throw new Error('Request Failed');
  }

  return response.json();
};
```

but i'm not sure if this is the right direction tho

**Reply by sthasambeg** [reply]
This blog by logrocket has a good solution for implementing intercepter like ability to fetch
https://blog.logrocket.com/axios-vs-fetch-best-http-requests/

> By default, fetch() doesn’t provide a way to intercept requests, but it’s not hard to come up with a workaround. You can overwrite the global fetch() method and define your own interceptor, like this:  
  
```js  
fetch = (originalFetch => {  
  return (...arguments) => {  
    const result = originalFetch.apply(this, arguments);  
      return result.then(console.log('Request was sent'));  
  };  
})(fetch);  
  
fetch('https://api.github.com/orgs/axios')  
  .then(response => response.json())  
  .then(data => {  
    console.log(data)   
  });  
```
